File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        
        vector<vector<int>>adj(numCourses);
        int n=prerequisites.size();
        for(auto it:prerequisites){
            adj[it[1]].push_back(it[0]);
        }

        vector<int>indeg(numCourses);
        for(int i=0;i<adj.size();i++){
            for(auto it:adj[i]){
                indeg[it]++;
            }
        }

        queue<int>q;
        for(int i=0;i<numCourses;i++){
            if(indeg[i]==0){
                q.push(i);
            }
        }

        vector<int>ans;
        while(!q.empty()){
            int top=q.front();
            q.pop();

            ans.push_back(top);
            for(auto it:adj[top]){
                indeg[it]--;
                if(indeg[it]==0){
                    q.push(it);
                }
            }
        }
        if(ans.size()==numCourses){
            return ans;
        }
        return {};
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int n=numCourses;
        List<List<Integer>> graph=new ArrayList<>();
        for(int i=0;i<n;i++){
            graph.add(new ArrayList<>());
        }

        int[] indeg=new int[numCourses];

        for(int[] pre:prerequisites){
            graph.get(pre[1]).add(pre[0]);
            indeg[pre[0]]++;
        }

        Queue<Integer>q=new LinkedList<>();
        for(int i=0;i<numCourses;i++){
            if(indeg[i]==0){
                q.offer(i);
            }
        }

        int[] ans=new int[numCourses];
        int idx=0;

        while(!q.isEmpty()){
            int node=q.poll();
            ans[idx++]=node;

            for(int nxt : graph.get(node)){
                indeg[nxt]--;
                if(indeg[nxt]==0){
                    q.offer(nxt);
                }
            }
        }

        if(idx!=numCourses){
            return new int[0];
        }

        return ans;
    }
}